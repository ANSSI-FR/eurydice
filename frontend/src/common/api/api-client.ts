import {
  camelCaseToSnakeCaseInterceptor,
  remoteUserHeaderForLoginRoute,
} from '@common/api/request-interceptors';
import {
  errorInterceptor,
  setUserFromResponseInterceptor,
  snakeCaseToCamelCaseInterceptor,
} from '@common/api/response-interceptors';
import axios from 'axios';

// Create an apiClient instance
export const initApiClient = () => {
  const API_URL = import.meta.env.VITE_API_URL ?? '/api/v1';
  const XSRF_COOKIE_NAME = import.meta.env.VITE_API_XSRF_COOKIE_NAME ?? 'eurydice_csrftoken';
  const XSRF_HEADER_NAME = import.meta.env.VITE_API_XSRF_HEADER_NAME ?? 'X-CSRFToken';

  const apiClient = axios.create({
    baseURL: API_URL,
    xsrfCookieName: XSRF_COOKIE_NAME,
    xsrfHeaderName: XSRF_HEADER_NAME,
    headers: {
      Accept: 'application/json',
    },
  });

  // This interceptor simulates reverse proxy header set for authentication
  // TODO: Find a way to do that in another place like traefik
  if (import.meta.env.VITE_USE_USER_REMOTE_INTERCEPTOR === 'true') {
    apiClient.interceptors.request.use((config) => remoteUserHeaderForLoginRoute(config));
  }

  // Request interceptor : converts camelCase request to snake_case
  apiClient.interceptors.request.use((config) => camelCaseToSnakeCaseInterceptor(config));

  // Response Interceptor : Set user from header
  apiClient.interceptors.response.use(
    (response) => {
      setUserFromResponseInterceptor(response);
      return response;
    },
    (error) => {
      setUserFromResponseInterceptor(error?.response);
      throw error;
    },
  );

  // Response Interceptor : Transforms snake_case to camelCase and intercepts errors
  apiClient.interceptors.response.use(
    (value) => snakeCaseToCamelCaseInterceptor(value),
    (error) => errorInterceptor(error),
  );

  return apiClient;
};

export const apiClient = initApiClient();
export default apiClient;
