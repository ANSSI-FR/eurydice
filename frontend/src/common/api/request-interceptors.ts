import { expandToSnakeCase, objectKeysToSnakeCase } from '@common/utils/case-functions';
import { type InternalAxiosRequestConfig } from 'axios';

export const camelCaseToSnakeCaseInterceptor = (config: InternalAxiosRequestConfig) => {
  // Convert object keys of the parameters from camelCase to snake_case
  if (config.params) {
    config.params = objectKeysToSnakeCase(config.params);

    if (config.params.expand) {
      config.params.expand = expandToSnakeCase(config.params.expand);
    }
  }
  // Convert object keys of the payload from camelCase to snake_case
  config.data = objectKeysToSnakeCase(config.data);
  return config;
};

export const remoteUserHeaderForLoginRoute = (config: InternalAxiosRequestConfig) => {
  if (config.url === '/user/login/') {
    config.headers['X-Remote-User'] = 'billmurray';
  }
  return config;
};
