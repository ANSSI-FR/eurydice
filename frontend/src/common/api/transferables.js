import request from "@common/utils/request";

export function listTransferables(params = {}, abortSignal) {
  return request({
    url: "/transferables/",
    signal: abortSignal,
    params,
  });
}

export function retrieveTransferable(transferableId, params = {}) {
  return request({
    url: `/transferables/${transferableId}/`,
    params,
  });
}
