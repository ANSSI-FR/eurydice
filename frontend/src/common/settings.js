module.exports = {
  title: "Eurydice",
  baseURL: "/api/v1",
  version: process.env.VUE_APP_VERSION || "development",
  releaseCycle: "Alpha",
  refreshIntervalInMs: 5 * 1000,
  serverDownIntervalInMs: 3 * 60 * 1000, // 3 min
  transferablesPerPage: 10,
};
