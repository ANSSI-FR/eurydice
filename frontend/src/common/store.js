import Vue from "vue";
import Vuex from "vuex";

Vue.use(Vuex);

export default new Vuex.Store({
  state: () => ({
    username: null,
    maintenance: false,
    serverDown: false,
    serverMetadata: null,
    ongoingTransfer: false,
  }),
  getters: {
    maintenance(state) {
      return state.maintenance;
    },
    serverDown(state) {
      return state.serverDown;
    },
    serverMetadata(state) {
      return state.serverMetadata;
    },
    ongoingTransfer(state) {
      return state.ongoingTransfer;
    },
  },
  mutations: {
    setUsername(state, newUsername) {
      state.username = newUsername;
    },
    setMaintenance(state, newMaintenance) {
      state.maintenance = newMaintenance;
    },
    setServerDown(state, serverDown) {
      state.serverDown = serverDown;
    },
    setServerMetadata(state, serverMetadata) {
      state.serverMetadata = serverMetadata;
    },
    setOngoingTransfer(state, ongoingTransfer) {
      state.ongoingTransfer = ongoingTransfer;
    },
  },
  actions: {},
  // NOTE: consider adding modules to this common store in main.js
  //       when adding origin/destination specific state/mutations
  modules: {},
});
