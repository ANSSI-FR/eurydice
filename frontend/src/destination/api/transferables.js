import request from "@common/utils/request";

// eslint-disable-next-line import/prefer-default-export
export function deleteTransferable(transferableId) {
  return request({
    url: `/transferables/${transferableId}/`,
    method: "DELETE",
  });
}
