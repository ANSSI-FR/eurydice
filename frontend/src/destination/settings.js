module.exports = {
  associationTokenWordCount: 18,
};
