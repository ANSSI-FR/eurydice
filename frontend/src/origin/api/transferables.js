import request from "@common/utils/request";
import { TRANSFERABLE_MAX_SIZE } from "@origin/constants";

export function cancelTransferable(transferableId) {
  return request({
    url: `/transferables/${transferableId}/`,
    method: "DELETE",
  });
}

export function createTransferable(file) {
  return request({
    url: "/transferables/",
    method: "POST",
    data: file,
    headers: {
      "Content-Type": "application/octet-stream",
      "Metadata-Name": encodeURI(file.name),
    },
    onUploadProgress: file.progressUpdater,
    signal: file.abortController.signal,
    timeout: 0,
  });
}

export function validateTransferable(file) {
  if (file?.size > TRANSFERABLE_MAX_SIZE) {
    const err = new Error("413 Content Too Large");
    err.response = {
      status: 413,
      data: {
        detail: `La taille du fichier dépasse le maximum autorisé (${TRANSFERABLE_MAX_SIZE})`,
      },
    };
    throw err;
  }
}
