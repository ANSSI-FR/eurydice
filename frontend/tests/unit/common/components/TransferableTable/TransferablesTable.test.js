import { listTransferables } from "@common/api/transferables";
import TransferablesTable from "@common/components/TransferableTable/main";
import { transferablesPerPage } from "@common/settings";
import { createLocalVue, mount, shallowMount } from "@vue/test-utils";
import Vuetify from "vuetify";
import store from "@/common/store";

jest.mock("@common/api/transferables", () => ({
  listTransferables: jest.fn(),
}));

// hardcode transferablesPerPage
// NOTE: we cannot use a variable here because jest.mock calls are executed before imports
// https://jestjs.io/docs/manual-mocks#using-with-es-module-imports
jest.mock("@common/settings", () => ({
  transferablesPerPage: 10,
}));

describe("TransferablesTable", () => {
  // vuetify instance is not used in shallowMount
  const localVue = createLocalVue();
  let vuetify;

  beforeEach(() => {
    vuetify = new Vuetify();
  });

  const PaginatorControlsStub = {
    template: "<div>PaginatorControlsStub</div>",
    data() {
      return {
        page: 1,
      };
    },
  };

  const dummyProps = {
    headers: [
      { text: "Sith", value: "isJedi" },
      { text: "Jedi", value: "isSith" },
    ],
    states: {
      ALIVE: {
        color: "success lighten-1",
        label: "Vivant",
        icon: "mdi-check",
      },
      DEAD: {
        color: "error lighten-1",
        label: "Mort",
        icon: "mdi-alert-circle",
      },
    },
  };

  // using jest.spyOn to override created lifecycle method because vue-test-utils
  // has no build in tools for that:
  // https://github.com/vuejs/vue-test-utils/issues/166
  jest.spyOn(TransferablesTable, "created").mockImplementation(jest.fn());

  it("renders correctly", async () => {
    const wrapper = mount(TransferablesTable, {
      localVue,
      vuetify,
      stubs: { PaginatorControls: PaginatorControlsStub },
      propsData: dummyProps,
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it.each`
    transferableCount | expectedNumPages
    ${0}              | ${0}
    ${1}              | ${1}
    ${10}             | ${1}
    ${11}             | ${2}
    ${20}             | ${2}
  `(
    `updates page count to $expectedNumPages with $transferableCount transferables`,
    async ({ transferableCount, expectedNumPages }) => {
      const wrapper = shallowMount(TransferablesTable, {
        propsData: dummyProps,
      });
      expect(wrapper.vm.numPages).toBe(0);

      const newTransferables = { count: transferableCount };

      await wrapper.setData({ transferables: newTransferables });

      expect(wrapper.vm.transferables.count).toBe(transferableCount);

      expect(wrapper.vm.numPages).toBe(expectedNumPages);
    }
  );

  it.each([true, false])(
    "getTransferableFrom queries the given URL when initialShowRefresh is %p",
    async (initialShowRefresh) => {
      const wrapper = shallowMount(TransferablesTable, {
        store,
        data() {
          return {
            transferables: {
              pages: { next: "supertoken" },
              showRefresh: initialShowRefresh,
            },
          };
        },
        propsData: dummyProps,
      });

      listTransferables.mockReturnValue({ newItems: true });

      await wrapper.vm.getTransferableFrom("next", false, false);

      expect(listTransferables).toHaveBeenCalledWith(
        {
          page: "supertoken",
          pageSize: transferablesPerPage,
        },
        expect.any(AbortSignal)
      );
      listTransferables.mockClear();

      expect(wrapper.vm.isLoading).toBe(false);
      expect(wrapper.vm.showRefresh).toBe(true);
    }
  );
  it("getTransferablesPage correctly queries the given page", async () => {
    const wrapper = mount(TransferablesTable, {
      localVue,
      vuetify,
      stubs: { PaginatorControls: PaginatorControlsStub },
      propsData: dummyProps,
      data() {
        return {
          transferables: {
            offset: transferablesPerPage * 4,
            pages: { current: "supertoken" },
          },
        };
      },
    });

    const expectedTransferables = { meAnd: "michael" };

    listTransferables.mockReturnValue(expectedTransferables);

    const requestedPage = 2;

    await wrapper.vm.getTransferablesPage(requestedPage);

    expect(listTransferables).toHaveBeenCalledWith({
      delta: -3,
      from: "supertoken",
      pageSize: transferablesPerPage,
    });
    listTransferables.mockClear();

    expect(wrapper.vm.isLoading).toBe(false);
    expect(wrapper.vm.transferables).toEqual(expectedTransferables);
  });
  it("getFreshFirstPage correctly queries the first page", async () => {
    const wrapper = mount(TransferablesTable, {
      localVue,
      vuetify,
      stubs: { PaginatorControls: PaginatorControlsStub },
      propsData: dummyProps,
      data() {
        return {
          showRefresh: true,
        };
      },
    });

    const expectedTransferables = { meAnd: "michael" };

    listTransferables.mockReturnValue(expectedTransferables);

    await wrapper.vm.getFreshFirstPage();

    expect(listTransferables).toHaveBeenCalledWith({
      pageSize: transferablesPerPage,
    });
    listTransferables.mockClear();

    expect(wrapper.vm.showRefresh).toBe(false);
    expect(wrapper.vm.isLoading).toBe(false);
    expect(wrapper.vm.transferables).toEqual(expectedTransferables);
    // If the page change does not come from the PaginatorControls component,
    // the TransferableTable component must change the page number manually
    expect(wrapper.vm.$refs.paginatorControls.page).toEqual(1);
  });
});
