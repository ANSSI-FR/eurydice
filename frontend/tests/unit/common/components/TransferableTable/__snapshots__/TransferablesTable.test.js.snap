// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransferablesTable renders correctly 1`] = `
<div>
  <div
    class="v-card v-sheet v-sheet--outlined theme--light"
  >
    <div
      class="v-data-table v-data-table--has-top v-data-table--has-bottom theme--light"
    >
      <header
        class="v-sheet theme--light v-toolbar v-toolbar--flat"
        style="height: 64px;"
      >
        <div
          class="v-toolbar__content"
          style="height: 64px;"
        >
          <div
            class="v-toolbar__title"
          >
            Transferts
          </div>
           
          <div
            class="spacer"
          />
           
          <transition-stub
            name="scale-transition"
          >
            <button
              class="v-btn v-btn--has-bg theme--light v-size--default primary"
              style="display: none;"
              type="button"
            >
              <span
                class="v-btn__content"
              >
                <span
                  aria-hidden="true"
                  class="v-icon notranslate v-icon--left theme--light"
                >
                  <svg
                    aria-hidden="true"
                    class="v-icon__svg"
                    role="img"
                    viewBox="0 0 24 24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M17.65,6.35C16.2,4.9 14.21,4 12,4A8,8 0 0,0 4,12A8,8 0 0,0 12,20C15.73,20 18.84,17.45 19.73,14H17.65C16.83,16.33 14.61,18 12,18A6,6 0 0,1 6,12A6,6 0 0,1 12,6C13.66,6 15.14,6.69 16.22,7.78L13,11H20V4L17.65,6.35Z"
                    />
                  </svg>
                </span>
                 Nouveaux fichiers
            
              </span>
            </button>
          </transition-stub>
        </div>
      </header>
      <div
        class="v-data-table__wrapper"
      >
        <table>
          <colgroup>
            <col
              class=""
            />
            <col
              class=""
            />
          </colgroup>
          <thead
            class="v-data-table-header"
          >
            <tr>
              <th
                aria-label="Sith"
                class="text-start"
                role="columnheader"
                scope="col"
              >
                <span>
                  Sith
                </span>
              </th>
              <th
                aria-label="Jedi"
                class="text-start"
                role="columnheader"
                scope="col"
              >
                <span>
                  Jedi
                </span>
              </th>
            </tr>
          </thead>
          <thead>
            <tr
              class="v-data-table__progress"
            >
              <th
                class="column"
                colspan="2"
              >
                <div
                  aria-valuemax="100"
                  aria-valuemin="0"
                  class="v-progress-linear v-progress-linear--absolute v-progress-linear--visible theme--light"
                  role="progressbar"
                  style="height: 4px;"
                >
                  <div
                    class="v-progress-linear__background primary"
                    style="opacity: 0.3; left: 0%; width: 100%;"
                  />
                  <div
                    class="v-progress-linear__buffer"
                  />
                  <transition-stub
                    name="fade-transition"
                  >
                    <div
                      class="v-progress-linear__indeterminate v-progress-linear__indeterminate--active"
                    >
                      <div
                        class="v-progress-linear__indeterminate long primary"
                      />
                      <div
                        class="v-progress-linear__indeterminate short primary"
                      />
                    </div>
                  </transition-stub>
                </div>
              </th>
            </tr>
          </thead>
          <tbody>
            <tr
              class="v-data-table__empty-wrapper"
            >
              <td
                colspan="2"
              >
                Loading items...
              </td>
            </tr>
          </tbody>
        </table>
      </div>
      <div
        loading="true"
        num-pages="0"
      >
        PaginatorControlsStub
      </div>
    </div>
  </div>
</div>
`;
